### PR TITLE
Improvements to Hand Constraint juicyness

### PR DESCRIPTION
**Pull request by provencher**
# Overview

Addition of useHybridHandRotation for HandConstraintPalmUp.
- This adds an additional check in HandConstraintPalmUp's hand validation function, which changes the Rotation behavior if useHybridHandRotation is enabled.

Added better behavior for LookAtTrackedObject.
- This introduces a branch in CalculateProjectedSafeZoneRay of the HandConstraint base class, which computes the SafeZone Direction ray according to the rotation of the SolverHandler.TargetTransform, rather than exclusively from the head gaze.

The end result can be seen [here](https://twitter.com/prvncher/status/1211795961510584327)

With useHybridHandRotation enabled:
The menu matches your hand rotation until your palm and head gaze reach a threshold angle, which activates a different rotation behavior that makes the menu look at the camera instead.

In other words, it flips open and closed 
gUBsKIae9V

### CODE REVIEW

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 1]
```diff
@@ -55,6 +55,14 @@ public float FlatHandThreshold
             set { flatHandThreshold = value; }
         }
 
+        [SerializeField]
+        [Tooltip("When using hybrid hand rotation, solver hand rotation until the menu is aligned with the gaze sufficiently")]
+        private bool useHybridHandRotation = false;
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: Do you think this would make sense as an enum, in case there are other styles like this in the future (or any you can already think of)?
Or does it make more sense as something in addition to other settings?

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 2]
```diff
@@ -55,6 +55,14 @@ public float FlatHandThreshold
...
+
+        [SerializeField]
+        [Tooltip("Angle between hand up and camera forward, below which the hand menu follows the gaze, if useHybridHandRotation is active.")]
+        private float hybridHandRotationThresholdAngle = 60f;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: Thanks for these new behaviors! Please also [update the documentation](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/PullRequests.html#documentation-should-be-added-in-the-same-pull-request-as-a-fix--feature) to describe the new fields. The Solvers Page, section [Hand Menu with HandConstraint PalmUp](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/README_Solver.html#hand-menu-with-handconstraint-and-handconstraintpalmup) would be a great place to describe the new behavior. 

Updating documentation along with the new behaviors helps ensure that the documentation for MRTK stays up to date.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 3]
```diff
@@ -94,13 +105,31 @@ protected override bool IsValidController(IMixedRealityController controller)
                             }
                         }
                     }
+
+                    // Check if the palm angle meets the prescribed threshold
+                    palmFacingThresholdMet = palmCameraAngle < facingThreshold;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: Interesting, I just had the thought that this could be made more general since currently the logic is "if the palm faces the camera, follow the camera, otherwise follow the tracked object". I think it could be more general by saying "if the tracked object faces the camera, follow the camera, otherwise follow the tracked object". Then it could work for things like motion controllers, other joints, or even following other game objects.

Not required for this PR, but was just interesting.

Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: I agree that it could work for more than just hands!

Really it's a matter of checking the SolverHandler.TargetTransform - that said, facing the camera for a palm means it's up direction aligns with the gaze vector. 
For a general purpose Solver, it's forward might be more useful. It's difficult to generalize that.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 2 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: I just updated the docs!

**Comment by provencher** [comment]
@julenka Thanks for your review! Let me know if I've addressed all your blocking concerns for this PR.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: An enum might help, but really that would need a system review of this constraint/solver, to ensure all required patterns are handled.  It would mean the implementation would need to change to ensure it is extensible by default.

Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: Thanks for chiming in @SimonDarksideJ! HandConstraints in general are a bit tacked on right now, as there is no reason for them to be Hand exclusive. I think a general system review would be merited.

Beyond that, there are other conditions I can think of that would affect the rotation mode, or even the overall Solver behavior.

For example, I've written some code that involves keeping a hand menu facing the user, while it's FaceCamera vector is lined up with the GazeVector. 
In other words, even if the palm is no longer facing the camera, the menu will keep facing the camera as long as the user is gazing it. This has the benefit of keep menus stable, even when the menu is fairly large.

Implementing a feature like that would require some rethinking about how hands here.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs [hunk 4]
```diff
@@ -416,8 +416,15 @@ private static Ray CalculateProjectedSafeZoneRay(Vector3 origin, IMixedRealityCo
                 default:
                 case SolverSafeZone.UlnarSide:
                     {
-                        direction = Vector3.Cross(CameraCache.Main.transform.forward, Vector3.up);
-                        direction = IsPalmFacingCamera(hand) ? direction : -direction;
+                        if (rotationBehavior == SolverRotationBehavior.LookAtTrackedObject)
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs]: Could you explain why we are setting direction to transform.right if the behavior is LookAtTrackedObject? Is it a fix for another issue you found?

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 5]
```diff
@@ -15,17 +16,18 @@ public class HandConstraintPalmUp : HandConstraint
...
         [Tooltip("The angle (in degrees) of the cone between the palm's up and camera's forward have to match. Only supported by IMixedRealityHand controllers.")]
         [Range(0.0f, 90.0f)]
-        private float facingThreshold = 80.0f;
+        private float facingCameraTrackingThreshold = 80.0f;
 
         /// <summary>
         /// The angle (in degrees) of the cone between the palm's up and camera's forward have to match. Only supported by <see cref="Microsoft.MixedReality.Toolkit.Input.IMixedRealityHand"/> controllers.
         /// </summary>
-        public float FacingThreshold
+        public float FacingCameraTrackingThreshold
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: I believe this is actually a [breaking change](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/BreakingChanges.html) since it is a change to a public property in a sdk asset. The new naming does seem more descriptive which is a positive. A drawback is that people may get compile errors if they update. This might be a case [where the breaking change adds value but can be written in a way that isn't breaking](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/BreakingChanges.html#the-breaking-change-adds-value-but-could-be-written-in-a-way-that-isnt-breaking).

What are your thoughts on the value vs drawbacks of this rename?

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 6]
```diff
@@ -55,6 +57,14 @@ public float FlatHandThreshold
             set { flatHandThreshold = value; }
         }
 
+        [SerializeField]
+        [Tooltip("With this active, solver will follow hand rotation until the menu is sufficiently aligned with the gaze, at which point it faces the camera.")]
+        private bool followHandUntilFacingCamera = false;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: It would be good to make these two fields configurable at runtime, [following the principle 'support configuring in editor and at runtime](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#support-configuring-components-both-in-editor-and-at-run-time)

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 7]
```diff
@@ -55,6 +57,14 @@ public float FlatHandThreshold
...
+
+        [SerializeField]
+        [Tooltip("Angle (in degrees) between hand up and camera forward, below which the hand menu follows the gaze, if followHandUntilFacingCamera is active.")]
+        private float followHandCameraFacingThresholdAngle = 60f;
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: It would be [good to write a test](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#prioritize-writing-tests) to ensure this behavior doesn't regress from somebody else's future changes. [Here is some information for how to write an run tests](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/UnitTests.html).

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs (hunk 4 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs]: Certainly!

So the the reason I changed this is because the follow tracked object rotation behavior sets the position at a constant offset relative to the camera to object vector. 

In my opinion, LookAtTrackedObject should be positioned relative the object's rotation, regardless of the camera's position. This is the meat of this PR to be honest.

However, I realize this may in fact change the behavior MRTK user have come to expect, and so perhaps we should make this change as part of a new rotation behavior enum.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 5 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: My mistake! I changed the private value and add the formerly serialized attribute to persist user's previous values in existing prefabs.

To be honest, I don't many folks would be changing this value at runtime via a script. Nothing in the MRTK does so, so it would be entirely in user app level code. If you want me to revert it, I'll do so, but I doubt many folks will notice.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 6 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: That makes sense! I can expose them as public properties.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 7 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: Hmm... I don't really know how to write a test that might make sense in this case.
Do you think writing a test is a blocking part of this pr?

**Comment by provencher** [comment]
@julenka Thanks for your second review pass! I added the both of you as collaborators.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 5 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: I think the serialized field can be modified but for the public property make it twice

```
public float FacingCameraTrackingThreshold
{
}
[System.Obsolete("Use FacingCameraTrackingThreshold property instead")
public float FacingThreshold
{
}

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 8]
```diff
@@ -15,17 +16,18 @@ public class HandConstraintPalmUp : HandConstraint
...
         /// <summary>
         /// The angle (in degrees) of the cone between the palm's up and camera's forward have to match. Only supported by <see cref="Microsoft.MixedReality.Toolkit.Input.IMixedRealityHand"/> controllers.
         /// </summary>
-        public float FacingThreshold
+        public float FacingCameraTrackingThreshold
         {
-            get { return facingThreshold; }
-            set { facingThreshold = value; }
+            get { return facingCameraTrackingThreshold; }
+            set { facingCameraTrackingThreshold = value; }
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: nit: while here can you switch to using =>
i.e
```
get => facingCameraTrackingThreshold;
set => facingCameraTrackingThreshold = value;
```

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 7 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: I think a basic test would be test starting below the threshold, crossing it, and then crossing back below the threshold. And between these changes validate that the solver positions the object at the expected different places (in this case hand vs camera following)

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs [hunk 9]
```diff
@@ -94,13 +107,31 @@ protected override bool IsValidController(IMixedRealityController controller)
                             }
                         }
                     }
+
+                    // Check if the palm angle meets the prescribed threshold
+                    palmFacingThresholdMet = palmCameraAngle < facingCameraTrackingThreshold;
+
+                    // If using hybrid hand rotation, we proceed with additional checks
+                    if (followHandUntilFacingCamera && palmFacingThresholdMet)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: nit: Reminder once you create the public properties, use those here and not the private instances

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs [hunk 10]
```diff
@@ -416,8 +416,15 @@ private static Ray CalculateProjectedSafeZoneRay(Vector3 origin, IMixedRealityCo
...
+                        {
+                            direction = targetTransform.right;
+                        }
+                        else
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs]: i feel like here and below should be 
`else if (rotationBehavior == SolverRotationBehavior.LookAtMainCamera)`
?
not sure yet what result should be when SolverRotationBehavior.None?

**Comment by provencher** [comment]
Thanks for the review pass @Troy-Ferrell, I'll try and make some of these additional changes soon. Hopefully some time this week.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs (hunk 10 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs]: @Troy-Ferrell I added another enum here called SolverOffsetBehavior for clarity. There is no none option here so this is no longer an issue. I believe the additional enum adds both clarity and options for users.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 8 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: I can but none of the other props in the class use this.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 9 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: Done.

**Comment by provencher** [comment]
@Troy-Ferrell I updated the code based on the feedback. Can try working on a test.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 7 above)
Comment (provencher) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: If someone writes a test for the hand constraint class, I'll happily add a function to test my changes. I just don't have the time to write up an entire test class for it right now.

File: Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs (hunk 8 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs]: sorry I meant for all of the props

**Comment by provencher** [comment]
Ok @Troy-Ferrell I added the test, and changed the props as you asked.
That said, When I run any playmode test, after the test runs, I get  bunch of errors.

My test seems to run as expected however, so if these playmode tests work for you guys, this new one should behave as expected.

image

**Comment by Troy-Ferrell** [comment]
> Ok @Troy-Ferrell I added the test, and changed the props as you asked.
> That said, When I run any playmode test, after the test runs, I get bunch of errors.
> 
> My test seems to run as expected however, so if these playmode tests work for you guys, this new one should behave as expected.
> 
> image

That is odd. Nothing looks out-of-ordinary with the test. If it passes in PR validation, then it's fine. 

Things look good. Only last key piece missing is the obsolete property
```c#
[System.Obsolete("Use FacingCameraTrackingThreshold property instead")
public float FacingThreshold
{
...
}
```

Add that in and we can kick off PR validation

**Comment by provencher** [comment]
Cheers! Added the prop as requested.

**Comment by provencher** [comment]
@keveleigh it seems that tests unrelated to my PR have failed. Would re-running solve this?
Also what are the next steps to get this merged in?

**Comment by keveleigh** [comment]
@provencher Just triggered a re-run, and filed the failing test at https://github.com/microsoft/MixedRealityToolkit-Unity/issues/7160 so we can make it more consistent.

I'll take a look today about merging this in!
